[CL-3116] Use block_end_at for blocked? status and realign/adjust user scopes

jinjagit opened the pull request:
# Changes
Set and write a `user.block_end_at` date-time in the User record when a user is blocked, and use this value to calculate `blocked?` and `scope: blocked`. Previously, we calculated this dynamically, depending on value of the setting `user_blocking['duration']`, which could lead to users flip-flopping between blocked/unblocked status if duration was changed in certain ways.

**Also includes:**
- Change model and controller to implement this change.
- Add new column to template serializers.
- Update copy in AdminHQ to reflect this change.
- Fix tests broken by this change.
- Include `block_start_at` & `block_end_at` in logged Activity.
- Realign `scope: active` with new `active?` definition in User model (I had overlooked this)
- Change use of `User.active` scope to `User.active_and_blocked` where this seems correct (see 'Assumptions/decisions' section, for more details). Please challenge my decisions to make / not make these changes, if you disagree.

# Assumptions / decisions

I have not changed the multiple uses of `User.active` (to `active_and_blocked`) in [stats_users_controller.rb](https://github.com/CitizenLabDotCo/citizenlab/blob/master/back/app/controllers/web_api/v1/stats_users_controller.rb)  because I imagine we do not want to include blocked users in these stats.

I'm asking Adrien about the `User.active` scope in [r_scores_controller.rb](https://github.com/CitizenLabDotCo/citizenlab/blob/master/back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/r_scores_controller.rb)

We don't want to notify blocked users of official feedbacks, e.g. `OfficialFeedbackOnCommentedInitiative` [see Slack thread](https://citizenlabco.slack.com/archives/C04SGJJR63X/p1678823402985459).

# Changelog
## Technical
- [CL-3116] Use new user.block_end_at date in user blocking feature (feature in development)


[CL-3116]: https://citizenlab.atlassian.net/browse/CL-3116?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jinjagit committed:
[CL-3116] Add column and migrate

jinjagit committed:
[CL-3116] Ser block_end_at when user blocked

jinjagit committed:
[CL-3116] Use block_end_at to set blocked? and in blocked scope

jinjagit committed:
[CL-3116] Update copy for adminHQ block duration setting

jinjagit committed:
[CL-3116] Set user_blocking duration minimum to 1 day

jinjagit committed:
[CL-3116] Simplify blocked? & scope: blocked by using only block_end_at

jinjagit committed:
[CL-3116] Fix policy specs

jinjagit committed:
[CL-3116] Fix accpetance tests

jinjagit committed:
[CL-3116] Remove unnecessary shared_context

jinjagit committed:
[CL-3116] Add block_start_at & block_end_at to block user Activity

jinjagit committed:
[CL-3116] Fix tests in after_fx_user_spec

jinjagit committed:
[CL-3116] Realign scope: active with new active? definition

jinjagit commented on back/app/controllers/web_api/v1/users_controller.rb:
```diff
@@ -13,7 +13,7 @@ def index
     @users = @users.search_by_all(params[:search]) if params[:search].present?
 
-    @users = @users.active unless params[:include_inactive]
+    @users = @users.active_and_blocked unless params[:include_inactive]
```
This may look weird, but it is intended.

Our back office `/admin/users` "Registered Users" view will now include users defined as active _and_ also  blocked users.

This makes sense, as it gives a central list of not-blocked _and_ blocked users that also excludes inactive users (not completed registration or invite pending).

I have attempted to keep all affected/new scopes logical and explicit in their naming.

jinjagit replied:
Now using `:registered` name for this scope.

jinjagit committed:
[CL-3116] Include new column in template serializers

jinjagit committed:
[CL-3116] Use new active_and_blocked scope in index_xlsx action

jinjagit committed:
[CL-3116] Replace active scope with active_and_blocked scope where appropriate

jinjagit commented:
@adessy I was not sure whether to change the use of the `User.active` scope to the new `active_and_blocked` scope, introduced in this PR, in [r_scores-controller.rb](https://github.com/CitizenLabDotCo/citizenlab/blob/master/back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/r_scores_controller.rb#L20).

I guessed not, as it appears it might be part of the stats functionality, and I have also guessed we don't want to include blocked users in stats, but I'm not sure about either of those assumptions.

What do you think?

jinjagit commented:
@adessy & @sebastienhoorens I asked for your reviews, mainly to get your opinions on where I have changed the use of `User.active` scope to a new `User.active_and_blocked` scope (a number of places).

Sometimes I am pretty confident these changes are appropriate, other times I feel I am guessing somewhat.
Also, if you can think of anywhere I should be changing the existing `User.active` scope _to_ `User.active_and_blocked`, please let me know.

Initially, it probably won't make much difference in any case, since we won't have many blocked users, but as time goes by the number will likely increase.

Thanks!

jinjagit committed:
[CL-3116] Don't send official feedback to blocked users

jinjagit committed:
Merge branch 'master' into CL-3116-Use-block-end-at-for-blocked-status

adessy commented:
> @adessy I was not sure whether to change the use of the `User.active` scope to the new `active_and_blocked` scope, introduced in this PR, in [r_scores-controller.rb](https://github.com/CitizenLabDotCo/citizenlab/blob/master/back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/r_scores_controller.rb?rgh-link-date=2023-03-14T20%3A05%3A58Z#L20).
> 
> I guessed not, as it appears it might be part of the stats functionality, and I have also guessed we don't want to include blocked users in stats, but I'm not sure about either of those assumptions.
> 
> What do you think?

I think it's more of a product question. However, if we don't change the user counts that are shown in the representativeness dashboard, we should not touch to the R-score either for the sake of consistency.

alexander-cit commented on back/app/models/user.rb:
```diff
@@ -235,13 +236,14 @@ def from_token_request(request)
+  scope :active_and_blocked, lambda {
+    where("registration_completed_at IS NOT NULL AND invite_status is distinct from 'pending'")
   }
+  scope :not_blocked, -> { where(block_end_at: nil).or(where('? > block_end_at', Time.zone.now)) }
```
I think `where('? > block_end_at', Time.zone.now)` does not include nil values, so you could simplify it to

```suggestion
  scope :not_blocked, -> { where('? > block_end_at', Time.zone.now) }
```

jinjagit replied:
For `:blocked`, yes, for `:not_blocked`, no.

We need to include records with `nil` values for `block_end_at` in the `:not_blocked` scope.

alexander-cit replied:
True, sorry for the confusion.

alexander-cit commented on back/spec/acceptance/users_spec.rb:
```diff
@@ -329,7 +329,27 @@
         end
 
-        example_request 'List all users includes user blocking related data' do
+        example 'List all users does not include inactive users', document: false do
```
I still think it's enough to have only 1-2 acceptance tests on a feature. And all other cases can be covered by model unit tests like here https://github.com/CitizenLabDotCo/citizenlab/pull/4203/files#diff-248e6600d9e8f3b06c088adf0f6aead2ccea0cde3078e7b2789a4b1e4e8a2e10R493

But up to you.

alexander-cit commented on back/app/models/user.rb:
```diff
@@ -235,13 +236,14 @@ def from_token_request(request)
+  scope :active, lambda {
+    where("registration_completed_at IS NOT NULL AND invite_status is distinct from 'pending'").not_blocked
+  }
+  scope :active_and_blocked, lambda {
```
I remember we were discussing it. But now I think that `active_and_blocked` is a bit confusing (partly because `active` is a bit vague). Maybe `registered` would be better? Or `completed_registration`?

alexander-cit commented on back/app/models/user.rb:
```diff
@@ -380,13 +382,11 @@ def active?
-    return nil unless blocked?
-
-    block_start_at + block_duration.days
+  def active_or_blocked?
```
Would `registered?` or `completed_registration?` work here?

alexander-cit replied:
And then

```ruby
  def active?
    registered? && !blocked?
  end
```

jinjagit replied:
Maybe. `active?` includes`!invite_pending?`, but I don't think you can have `registration_completed_at.present?` if you have `invite_pending?` anyway 🤔

jinjagit replied:
But I might include it anyway, to avoid some weird edge case we haven't though of. So we could have...
```
def registered?
  registration_completed_at.present? && !invite_pending?
end

jinjagit replied:
I thought I could get away with removing `!invite_pending` from the picture, but it seems it would require work that may not be useful in the light of planned changes by the methods squad. (See [Slack thread](https://citizenlabco.slack.com/archives/C016C2EHURY/p1678992484306909
))

alexander-cit replied:
I think it's the right decision to leave `!invite_pending`. Definitely, it was added for a reason.

jinjagit replied:
Changed approach and removed it. See [comment](https://github.com/CitizenLabDotCo/citizenlab/pull/4252#issuecomment-1474102081).

alexander-cit commented on back/app/models/user.rb:
```diff
@@ -235,13 +236,14 @@ def from_token_request(request)
-      .and(where(block_start_at: (
-        AppConfiguration.instance.settings('user_blocking', 'duration').days.ago..Time.zone.now)))
+  scope :active, lambda {
+    where("registration_completed_at IS NOT NULL AND invite_status is distinct from 'pending'").not_blocked
```
`registered.not_blocked`?

alexander-cit reviewed (commented):
It seems `_or_blocked` and `_and_blocked` are a bit hard to figure out. Also, sometimes we first check for blocked, and then for not blocked. Maybe different naming could simplify it?

jinjagit committed:
[CL-3116] Refactor to simplify User.active? and related scopes

jinjagit committed:
[CL-3116] Fix :active scope

jinjagit committed:
[CL-3116] Replace not invite_pending in registered? & scope

alexander-cit reviewed (approved):
Now I realize that `completely_registered` or `completed_registration` could be better, but if you like `register`, it's fine too.

jinjagit commented:
> Now I realize that `completely_registered` or `completed_registration` could be better, but if you like `register`, it's fine too.

I can change it. No problem. :+1: I'm going to chat with @jamesspeake about `active?` this morning, so maybe he will help us understand why we care so much about `!invite_pending`.

jinjagit committed:
[CL-3116] Use 'fully_registered', not 'registered'

jinjagit committed:
[CL-3116] Revert to 'registered', not 'fully_registered'

Following discussion with james Speake, we decided this name is,
and will be, less confusing. Also, that I should probably look into
removing the !invite_pending from the method & scope.

jinjagit committed:
Merge branch 'CL-2623-User-blocking-epic-i1' into CL-3116-Use-block-end-at-for-blocked-status

jinjagit committed:
[CL-3116] Fix broken test

jinjagit committed:
[CL-3116] Remove !invite_pending from registered? & related scope

It seems this was compensating for a flawed :invited_user factory
that produced user objects with invte_status: 'pending' and a
registration_completed_at: <date>. This also fixes that factory.

jinjagit committed:
[CL-3116] Remove redundant test

jinjagit committed:
[CL-3116] Add test of registered?

jinjagit committed:
[CL-3116] Remove some acceptance tests (where adequate model tests coverage)

jinjagit commented:
@alexander-cit FYI:

After discussing further with @jamesspeake, we decided that removing `!invite_pending` from the `registered?` method (and related scope) is probably a good idea. It seemed that it was simply compensating for a flawed `:invited_user` factory that create User objects with a date value for `registration_completed_at`.

Fixing this factory and removing the `!invite_pending` from the `registered?` method simplifies & clarifies things, and also makes the `registered?` naming unambiguous, as it no longer incorporates any other concept.

I also removed a few more acceptance tests where I have adequate coverage in model tests.
